### ISSUE TITLE
MismatchedTokenError caused by use of exec() function

### ISSUE THREAD

**Issue by ceridwen** [python-rope/rope#117]
The problem is that `ast` compiles both the exec statement and exec function into the same AST nodes, while patchedast.py assumes the `exec` statement.

``` python
>>> import ast
>>> ast.Exec
<class '_ast.Exec'>
>>> ast.Exec._fields
('body', 'globals', 'locals')
>>> ast.dump(ast.parse('exec "def f(): pass" in {}'))
"Module(body=[Exec(body=Str(s='def f(): pass'), globals=Dict(keys=[], values=[]), locals=None)])"
>>> ast.dump(ast.parse('exec("def f(): pass", {})'))
"Module(body=[Exec(body=Str(s='def f(): pass'), globals=Dict(keys=[], values=[]), locals=None)])"
```

From `patchedast.py`, lines 335-342.

``` python

    def _Exec(self, node):
        children = []
        children.extend(['exec', node.body])
        if node.globals:
            children.extend(['in', node.globals])
        if node.locals:
            children.extend([',', node.locals])
        self._handle(node, children)
```

I'm willing to write a patch and pull request, but what's the right way to rewrite _Exec?

**Comment by mcepl** [comment]
@aligrudi I have missed this one. Could you please show here some opinion, please?

**Comment by aligrudi** [comment]
ceridwen <notifications@github.com> wrote:
> The problem is that `ast` compiles both the exec statement and exec function into the same AST nodes, while patchedast.py assumes the `exec()` statement.
> 
> ```python
> >>> import ast
> >>> ast.Exec
> <class '_ast.Exec'>
> >>> ast.Exec._fields
> ('body', 'globals', 'locals')
> >>> ast.dump(ast.parse('exec "def f(): pass" in {}'))
> "Module(body=[Exec(body=Str(s='def f(): pass'), globals=Dict(keys=[], values=[]), locals=None)])"
> >>> ast.dump(ast.parse('exec("def f(): pass", {})'))
> "Module(body=[Exec(body=Str(s='def f(): pass'), globals=Dict(keys=[], values=[]), locals=None)])"
> ```
> 
> >From `patchedast.py`, lines 335-342.
> 
> ```python
> 
>     def _Exec(self, node):
>         children = []
>         children.extend(['exec', node.body])
>         if node.globals:
>             children.extend(['in', node.globals])
>         if node.locals:
>             children.extend([',', node.locals])
>         self._handle(node, children)
> ```

Does the second case cause an exception in rope?

> I'm willing to write a patch and pull request, but what's the right way to rewrite _Exec?

See _excepthandler() and how it uses semicolon_or_as_in_except
to handle a similar case.

	Ali

**Comment by lieryan** [comment]
Closing. This issue will be fixed with PR #372.
